scripts/buf.pl

```diff
@@ -40,64 +43,59 @@
 sub _filename { sprintf '%s/scrollbuffer', get_irssi_dir }
 
 sub upgrade {
-    my $fn = _filename;
-    my $old_umask = umask 0077;
-    open my $fh, q{>}, $fn or die "open $fn: $!";
-    umask $old_umask;
-
-    print $fh join("\0", map $_->{server}->{address} . $_->{name}, channels), "\n";
+    my $out = { suppress => join("\0", map $_->{server}->{address} . $_->{name}, channels) };
     for my $window (windows) {
-	next unless defined $window;
-	next if $window->{name} eq 'status';
-	my $view = $window->view;
-	my $line = $view->get_lines;
-	my $lines  = 0;
-	my $buf = '';
-	if (defined $line){
-	    {
-		$buf .= $line->get_text(1) . "\n";
-		$line = $line->next;
-		$lines++;
-		redo if defined $line;
-	    }
-	}
-	printf $fh "%s:%s\n%s", $window->{refnum}, $lines, $buf;
+        next unless defined $window;
+        next if $window->{name} eq 'status';
+        my $view = $window->view;
+        my $line = $view->get_lines;
+        my $lines  = 0;
+        my $buf = '';
+        my $output;
+        if (defined $line) {
+            {
+              push @$output, { level => $line->{info}{level}, data => $line->get_text(1) };
+              $line = $line->next;
+              redo if defined $line;
+            }
+        }
+        push @{$out->{windows}}, { refnum => $window->{refnum}, lines => $output };
     }
-    close $fh;
+    my $old_umask = umask 0077;
+    store($out, _filename) or die $!;
```

Comment:
I can't find any mention in the docs that Storable sets $!. I think you would have to die with a generic error string ("storing failed")